### PR TITLE
Weight aggregation with data volumes

### PR DESCRIPTION
**Pull request by bowni**
- Fixes issue #15
- Between each `.fit` iteration, the aggregation of weights is now weighted with the data volumes of each node
- For that, `scenario.amounts_per_node` is now passed in `fl_training.compute_test_score`

### CODE REVIEW

File: fl_training.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ def compute_test_score_with_scenario(scenario, is_save_fig=False):
         scenario.x_test,
         scenario.y_test,
         scenario.is_early_stopping,
+        scenario.amounts_per_node,
```
Comment (RomainGoussault) [inline fl_training.py]: At this point, I would pass the whole scenario object to the `compute_test_score_with_scenario()` method

File: fl_training.py [hunk 2]
```diff
@@ -161,18 +163,24 @@ def compute_test_score(
                 print("   First epoch, generate model from scratch")
 
             else:
-                # Aggregating phase : averaging the weights
+                # Aggregating phase: averaging the weights
                 print("   Aggregating models weights to build a new model")
-                weights = [model.get_weights() for model in model_list]
+                weights_per_model = [model.get_weights() for model in model_list]
+                weights_per_layer = list(zip(*weights_per_model))
```
Comment (RomainGoussault) [inline fl_training.py]: It is way clearer 👍

File: fl_training.py [hunk 3]
```diff
@@ -161,18 +163,24 @@ def compute_test_score(
...
                 new_weights = list()
 
-                for weights_list_tuple in zip(*weights):  # TODO : make this clearer
-                    new_weights.append(
-                        [
-                            np.array(weights_).mean(axis=0)
-                            for weights_ in zip(*weights_list_tuple)
-                        ]
-                    )
+                # OLD:
```
Comment (RomainGoussault) [inline fl_training.py]: remove commented code

File: fl_training.py [hunk 4]
```diff
@@ -121,6 +122,7 @@ def compute_test_score(
     x_test,
     y_test,
     is_early_stopping=True,
+    aggregation_weights=None,
```
Comment (RomainGoussault) [inline fl_training.py]: I think we should have a param in the scenario class to choose which aggregation we want to perform (for now it will only be: weighted by data size or weighted by nothing)
It will also be easier to test the impact of the weighted average, so we can be sure that weighted average is better than no weight.

File: fl_training.py (hunk 1 above)
Comment (bowni) [reply fl_training.py]: You mean to `compute_test_score()`?

Comment (RomainGoussault) [reply fl_training.py]: Yes you are right, I meant `compute_test_score()`

Comment (bowni) [reply fl_training.py]: We can't do that simply. Because we need a `compute_test_score()` independent from scenarios, to be used in contributivity measure algorithms when needing to launch federated learning on an arbitrary coalition.

File: fl_training.py (hunk 4 above)
Comment (bowni) [reply fl_training.py]: Ok, done in 6b0ec01.

File: fl_training.py (hunk 3 above)
Comment (bowni) [reply fl_training.py]: Ok, done in 1cce0b6.

File: fl_training.py (hunk 1 above)
Comment (RomainGoussault) [reply fl_training.py]: Ah yes you are right !

**Comment by RomainGoussault** [comment]
I am currently running some scenario to see the performance improvement on heavy imbalanded data distribution between partners.

**Comment by bowni** [comment]
@RomainGoussault I just rebased it with master with the recent merge of the pre-processing refacto and the central datasets renaming.

File: fl_training.py [hunk 5]
```diff
@@ -142,6 +176,13 @@ def compute_test_score(
         global_val_acc = []
         global_val_loss = []
 
+        # Create list of weights for aggregation steps
+        aggregation_weights = []
```
Comment (RomainGoussault) [inline fl_training.py]: I think the `aggregation_weights`  should sum to 1:

- uniform:` [1/nodes_count] * nodes_count`
- data-volume: `[len(node.x_train) for node in node_list] / total_data_size_in_nodes`

et un petit `assert(np.sum(aggregation_weighting ) == 1) `à la fin

**Comment by RomainGoussault** [comment]
First test on 0.95 / 0.05 distribution, there is not much difference. I am trying on 0.99 / 0.01 distribution now.
image

**Comment by Thomas-Galtier** [comment]
Ben on a de meilleurs resultats plus de 50% du temps non?

**Comment by RomainGoussault** [comment]
5 repeats with new data distribution. 
image

File: fl_training.py (hunk 5 above)
Comment (bowni) [reply fl_training.py]: Ok, doing that now.

Comment (bowni) [reply fl_training.py]: Done in cb625f7.

**Comment by RomainGoussault** [comment]
image
(nothing is like uniform but normalized, it was run before the normalization commit https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/80/commits/cb625f7a9e486189d13c92f4d0c188ff9ee7f032)
